### PR TITLE
add functionality and tests for classifier-free guidance

### PR DESCRIPTION
**Pull request by glenliu21**
## Motivation

Classifier-free guidance is a method that's commonly used in diffusion models to enhance adherence to provided conditions during generation at minimal cost. More recently, however, it has been shown to be an effective tool in [conditioning text generation](https://arxiv.org/abs/2306.17806) and even [enhancing audio quality in voice AI](https://github.com/resemble-ai/chatterbox). Support for it has landed in [Huggingface-Transformers](https://github.com/huggingface/transformers/pull/24654/commits/ee21d2549cc886f63d6e43b83de5cfbe7a6c385a), and this pull-request does the same for SGLang.

To see an example, consider this snippet of code (the prompts are taken from the Huggingface CFG PR):
```
llm = sgl.Engine(model_path="meta-llama/Meta-Llama-3.1-8B-Instruct")

prompt = "Today, a dragon flew over Paris, France,"

cfg_texts = [
    "A very sad event happened,",
    "A very happy event happened,",
]

cfg_params = [
    {
        "cfg_text": cfg_text,
        "cfg_weight": 1.0,
    }
    for cfg_text in cfg_texts
]

sampling_params = {
    "temperature": 0.3,
    "top_p": 1.0,
    "min_p": 0.05,
}

outputs = await llm.async_generate(
    [prompt] * len(cfg_texts),
    sampling_params,
    cfg_params=cfg_params,
)

for cfg_param, output in zip(cfg_params, outputs):
    print(f"\nCFG params: {cfg_param}")
    print(f"Generated text: {output['text']}")
```

Running this gives us the output:
```
CFG params: {'cfg_text': 'A very sad event happened,', 'cfg_weight': 1.0, 'cfg_input_ids': [128000, 32, 1633, 12703, 1567, 7077, 11]}
Generated text:  soaring above the Eiffel Tower and leaving a trail of magic in its wake. Witnesses reported seeing glittering sparks and a faint aura of wonder as the dragon banked and turned above the city.
"I was walking to the café when I saw it," said local resident Colette Dupont. "At first, I thought it was a plane or a bird, but then I saw the scales and the wings... it was a real, honest-to-goodness dragon!"
The dragon, which appeared to be a magnificent creature with scales that shone like gold and bronze, flew over the city for several minutes before disappearing into the clouds.
"I

CFG params: {'cfg_text': 'A very happy event happened,', 'cfg_weight': 1.0, 'cfg_input_ids': [128000, 32, 1633, 6380, 1567, 7077, 11]}
Generated text:  causing panic among Parisians. Witnesses described the dragon as large, with scales that shone like gold in the sunlight. It flew over the Eiffel Tower, leaving a trail of smoke and fire in its wake.
The dragon was seen flying over the city for several minutes before disappearing into the clouds. Authorities have confirmed that the dragon was real, but they have not yet determined its origin or intentions.
"I was walking home from work when I saw the dragon flying overhead," said one witness. "It was enormous, with wings that stretched from one side of the street to the other. I was terrified."
Another witness described the dragon's
```

Notice that the first output is decidedly more positive than the second one. The first one uses phrases like "trail of magic", "glittering sparks", and "faint aura of wonder", while the second output starts with the phrase "causing panic among Parisians" and uses words like "smoke", "fire", and "terrified". Note that because CFG utilizes negative-prompting, what we are seeing is the model "move away" from the provided `cfg_text`s.

## Modifications

Implementation-wise, CFG requests follow the below pipeline through SRT:
- CFG metadata is added onto requests before they are tokenized.
- Once the Scheduler sees a CFG request, it creates that request's unconditioned pair.
- From there, the Scheduler ensures that a CFG request is ran in a batch if and only if its unconditioned pair is as well.
- The TpWorker then applies logit biasing if it receives CFG requests in its batch according to the formula `logits = logits_cond + cfg_weight * (logits_cond - logits_uncond)`. Note that this specific formula may appear different from the standard formula used in CFG (ie. `out = self.guidance_scale * (scores - unconditional_logits) + unconditional_logits`, from HF), but they are equivalent if one sets `self.guidance_scale = 1 + cfg_weight`.
- Lastly, the Scheduler removes the unconditioned pair of CFG requests from the result before sending it to the detokenizer manager.

## Accuracy Tests

High-level, end-to-end tests have been added in `test-cfg.py`, and the outputs of the script serve as evidence for the effectiveness of CFG.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).

### CODE REVIEW

File: python/sglang/srt/managers/schedule_policy.py [hunk 1]
```diff
@@ -567,3 +567,48 @@ def add_one_req(self, req: Req, has_chunked_req: bool):
...
+
+            # Update snapshot
+            snapshot["rem_total_tokens"] -= total_tokens
+            snapshot["rem_input_tokens"] -= input_tokens
+            snapshot["cur_rem_tokens"] -= input_tokens
+            if snapshot["rem_chunk_tokens"] is not None:
+                snapshot["rem_chunk_tokens"] -= input_tokens
+
+                if snapshot["rem_chunk_tokens"] < 0 and len(self.can_run_list) == 0:
+                    raise RuntimeError("CFG requests cannot be chunk prefilled.")
```
[the remark this answers is not in the document]
Comment (glenliu21) [reply python/sglang/srt/managers/schedule_policy.py:723]: This `RuntimeError` addresses a special case where a request with CFG enabled is too large to fit in a batch even by itself. Because CFG requests can't be chunk-prefilled, this should throw an error, or else such a request would endlessly attempt to schedule.

**Comment by mickqian** [comment]
Nice work!
cc @merrymercy @zhyncs @Ying1123

**Comment by mickqian** [comment]
BTW,  is `logits = logits_cond + cfg_weight * (logits_cond - logits_uncond)` supposed to be `logits = logits_uncond + cfg_weight * (logits_cond - logits_uncond)` ?

**Comment by glenliu21** [comment]
@mickqian the two equations are actually (essentially) equivalent. That is, `logits = logits_cond + cfg_weight1 * (logits_cond - logits_uncond)` is the same thing as `logits = logits_uncond + cfg_weight2 * (logits_cond - logits_uncond)`, if we let `cfg_weight2 = 1 + cfg_weight1`. I am more used to setting the logits using the first equation, however, I'm happy to change it to the second if you or others think that is clearer.
